### PR TITLE
User story5

### PR DESCRIPTION
**Pull request by NicolaiKopka**

### CODE REVIEW

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/main_page/MainPageController.java [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
 import com.github.NicolaiKopka.dto.StreamingStatusDTO;
+import com.github.NicolaiKopka.dto.UserInfoDTO;
 import lombok.RequiredArgsConstructor;
 import org.apache.oltu.oauth2.common.exception.OAuthProblemException;
 import org.apache.oltu.oauth2.common.exception.OAuthSystemException;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.*;
 
+import java.security.Principal;
```
Comment (gossie) [inline backend/src/main/java/com/github/NicolaiKopka/web_page_services/main_page/MainPageController.java]: This import isn't used, is it?

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserController.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+@RestController
+@RequestMapping("/api/soundtracker/user")
+@RequiredArgsConstructor
+public class UserController {
+
+    private final UserService userService;
+
+    @GetMapping()
+    public ResponseEntity<UserInfoDTO> getUserInfo(Principal principal) {
+        return ResponseEntity.ok(userService.getUserInfo(principal.getName()));
```
Comment (gossie) [inline backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserController.java]: You don't need the `ResponseEntity` here. `ResponseEntity` only makes sense, if you want to return different status depending on the logic.
Just return the `UserInfoDTO`.

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserService.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+public class UserService {
+
+    private final MyUserRepo myUserRepo;
+    private final UserFavoritesRepo userFavoritesRepo;
+
+    public UserInfoDTO getUserInfo(String username) {
+
+        MyUser user = myUserRepo.findByUsername(username).orElseThrow();
+        UserFavoritesSaveObject favoritesObject = userFavoritesRepo.findByUserId(user.getId()).orElseThrow();
+        return null;
```
Comment (gossie) [inline backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserService.java:22]: This probably shouldn't be null.

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java [hunk 4]
```diff
@@ -30,23 +30,36 @@ public Collection<Movie> getAllFavoriteMoviesFromDbByUser(String username) {
...
-            // TODO any way to make return List mutable?
-            List<Integer> idList = new ArrayList<>(userFavorites.getMovieIds().stream().toList());
-            idList.add(movieId);
-            userFavorites.setMovieIds(idList);
+            List<Integer> movieIds = new ArrayList<>(List.copyOf(userFavorites.getMovieIds()));
+            if(movieIds.contains(movieId)) {
+                throw new IllegalArgumentException("Movie already in list");
+            }
+            movieIds.add(movieId);
+            userFavorites.setMovieIds(movieIds);
```
Comment (gossie) [inline backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java]: I remember that we talked about that. What I ment was something like that.
Here:
```java
userFavorites.addMovieId(movieId);
```
And in the `UserFavoritesSaveObject `:
```
public UserFavoritesSaveObject {
    
    private List<Integer> movieIds = new ArrayList<>();

    ...

    public void addMovieId(Integer movieId) {
        if(movieIds.contains(movieId)) {
            throw new IllegalArgumentException("Movie already in list");
        }
        movieIds.add(movieId);
    }
}
```

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java [hunk 5]
```diff
@@ -30,23 +30,36 @@ public Collection<Movie> getAllFavoriteMoviesFromDbByUser(String username) {
...
+    }
 
+    public UserFavoritesSaveObject deleteMovieFromFavorites(int movieId, String username) {
+
+        MyUser user = myUserRepo.findByUsername(username).orElseThrow();
+        UserFavoritesSaveObject favoritesObject = userFavoritesRepo.findByUserId(user.getId()).orElseThrow();
+        List<Integer> idList = new ArrayList<>(List.copyOf(favoritesObject.getMovieIds()));
+        idList.remove(Integer.valueOf(movieId));
+        favoritesObject.setMovieIds(idList);
+        return userFavoritesRepo.save(favoritesObject);
```
Comment (gossie) [inline backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java]: I would rewrite this method, so it returns an `Optional`. That way you can use the `ResponseEntity.of`-method.
```java
public Optional<UserFavoritesSaveObject> deleteMovieFromFavorites(int movieId, String username) {
    return myUserRepo.findByUsername(username)
            .flatMap(user -> userFavoritesRepo.findByUserId(user.getId()))
            .map(favoritesObject -> favoritesObject.deleteMovieId(movieId))
            .map(userFavoritesRepo::save)
}
```
And in `UserFavoritesSaveObject` you need a `deleteMovieId`-method:
```java
public void addMovieId(Integer movieId) {
    movieIds.removeIf(id -> Objects.equals(id, movieId));
    return this;
}
```

**Review by gossie — COMMENTED** [review]
I left some comments.

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserController.java (hunk 2 above)
Comment (NicolaiKopka) [reply backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserController.java]: UserController and UserService are not in use yet. Just had an idea but didn't finish. But also wanted to keep the idea for later

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserService.java (hunk 3 above)
Comment (NicolaiKopka) [reply backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_area/UserService.java:22]: UserController and UserService are not in use yet. Just had an idea but didn't finish. But also wanted to keep the idea for later

File: backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java (hunk 5 above)
Comment (NicolaiKopka) [reply backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java]: This I don't understand. I don't want to return the SaveObject in my Controller but a DTO. So I would need to build that in the Controller anyways.

Comment (NicolaiKopka) [reply backend/src/main/java/com/github/NicolaiKopka/web_page_services/user_pages/user_favorites/UserFavoritesService.java]: But I think the way I implemented it now works. Was that what you meant?

**Review by gossie — APPROVED** [review]

**Pull request merged**